include/rmm/mr/cnmem_memory_resource.hpp

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/detail/cnmem.h>
+#include "device_memory_resource.hpp"
+
+#include <cuda_runtime_api.h>
+#include <cassert>
+#include <exception>
+#include <iostream>
+#include <mutex>
+#include <set>
+
+namespace rmm {
+namespace mr {
+/**---------------------------------------------------------------------------*
+ * @brief `device_memory_resource` derived class that uses
+ * the cnmem pool sub-allocator for allocation/deallocation.
+ *---------------------------------------------------------------------------**/
+class cnmem_memory_resource final : public device_memory_resource {
```

Comment:
Do we want to expose cnmem to users this way? Perhaps this should be abstracted, first by naming it something more general, like `pool_memory_resource` or `suballocator_memory_resource`?  How do we ensure it's easy to choose different suballocator implementations? I don't want, for example, libcudf users to be explicitly passing cnmem allocators into libcudf APIs.

Reply:
My main thinking is that we likely want to experiment and provide other sub-allocators in the future. Therefore, I didn't want to chose a name like `pool_memory_resource` that would preclude having other pool allocator resource implementations. 

I imagine that most users will just be setting the global default resource once via `rmmInitialize`. We can keep the options at initialization abstract and just provide an option to enable "pool" mode, and then we can control which sub-allocator resource is actually used.

Reply:
We could perhaps put the `cnmem_memory_resource` in the `detail` namespace to indicate that it shouldn't be used directly.

Reply:
Ah, I thought this PR was proposing to do away with all rmm* interfaces.